### PR TITLE
New drivers and tests.

### PR DESCRIPTION
**Pull request by ptheberge**
PhpDriver, JsonDriver, YamlDriver and associated test files.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -17,7 +17,8 @@
         "doctrine/common": "~2.3",
         "doctrine/inflector": "~1.0",
         "leedavis81/drest-common": "dev-master",
-        "zendframework/zend-code": "~2.0"
+        "zendframework/zend-code": "~2.0",
+        "symfony/yaml": "2.*"
```
Comment (jdrich) [inline composer.json]: This should be a suggested repo as it isn't required if you're not using the YAML driver.

File: examples/application1/Entities/User.php [hunk 2]
```diff
@@ -15,27 +15,6 @@
 
 /**
  * User
- *
```
Comment (jdrich) [inline examples/application1/Entities/User.php]: You should not have removed this from the example.

File: examples/application1/Entities/User.php [hunk 3]
```diff
@@ -121,6 +100,7 @@ public function addPhoneNumber(PhoneNumber $phoneNumber)
      */
     public function patchUser(array $data)
     {
+        print_r($data);
```
Comment (jdrich) [inline examples/application1/Entities/User.php]: What?

File: examples/application1/public/index.php [hunk 4]
```diff
@@ -27,8 +27,8 @@
 
 $em = \Doctrine\ORM\EntityManager::create(array(
     'host' => 'localhost',
-    'user' => 'developer',
-    'password' => 'developer',
+    'user' => 'root',
```
Comment (jdrich) [inline examples/application1/public/index.php]: Why?

File: examples/application1/public/index.php [hunk 5]
```diff
@@ -46,20 +46,26 @@
 $drestConfig->setExposureDepth(3);
 $drestConfig->setExposeRequestOption(Configuration::EXPOSE_REQUEST_PARAM_GET, 'expose');
 $drestConfig->setMetadataCacheImpl(new \Doctrine\Common\Cache\ArrayCache());
-//$drestConfig->setDebugMode(true);
+$drestConfig->setDebugMode(true);
```
Comment (jdrich) [inline examples/application1/public/index.php]: Stop changing things!

File: examples/application1/public/index.php [hunk 6]
```diff
@@ -46,20 +46,26 @@
...
 $drestConfig->addPathsToConfigFiles($pathToEntities);
-
+$drestConfig->setAttribute('configFilePath', __DIR__);
```
Comment (jdrich) [inline examples/application1/public/index.php]: While we've introduced new functionality, we shouldn't break the existing annotation driven example.

File: examples/application1/public/index.php [hunk 7]
```diff
@@ -46,20 +46,26 @@
...
-
 $evm = new Event\Manager();
 //$evm->addEventListener(array('preServiceAction', 'postServiceAction', 'preRouting', 'postRouting', 'preDispatch', 'postDispatch'), new MyEvents\MyEvent());
 
 
 $emr = \Drest\EntityManagerRegistry::getSimpleManagerRegistry($em);
 $drestManager = \Drest\Manager::create($emr, $drestConfig, $evm);
+?>
 
+<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
```
Comment (jdrich) [inline examples/application1/public/index.php]: ?!?!?

File: src/Drest/Configuration.php [hunk 8]
```diff
@@ -84,6 +84,25 @@ public function __construct()
...
+     * Set an attribute for the driver configuration 
+     * @param string $attribute, string $value
+     */
+    public function setAttribute($attribute, $value) {
+        $this->_attributes[$attribute] = $value;
+    }
+
+    /**
+     * Get an attribute from the driver configuration
+     * @param string $attribute
```
Comment (jdrich) [inline src/Drest/Configuration.php]: Missing @return annotation.

File: src/Drest/Mapping/Driver/JsonDriver.php [hunk 9]
```diff
@@ -0,0 +1,203 @@
+<?php
+
+namespace Drest\Mapping\Driver;
+
+use Doctrine\Common\Annotations;
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/JsonDriver.php]: Why do you need annotations here?

File: src/Drest/Mapping/Driver/JsonDriver.php [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+/**
+ * The JsonDriver reads a configuration file (config.json) rather than utilizing annotations.
+ */
+class JsonDriver extends PhpDriver
+{
+
+    public function __construct()
+    {
+        $filename = self::$configuration_filepath . '\\' . self::$configuration_filename;
+        $json = json_decode(file_get_contents($filename), true);
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/JsonDriver.php]: What happens when the JSON is invalid?

File: src/Drest/Mapping/Driver/JsonDriver.php [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+use Drest\Mapping\Annotation;
+use Drest\Mapping;
+use Drest\Mapping\RouteMetaData;
+
+/**
+ * The JsonDriver reads a configuration file (config.json) rather than utilizing annotations.
+ */
+class JsonDriver extends PhpDriver
+{
+
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/JsonDriver.php]: Where are your docblocks?

File: src/Drest/Mapping/Driver/JsonDriver.php (hunk 10 above)
Comment (jdrich) [inline src/Drest/Mapping/Driver/JsonDriver.php]: You don't check to see if the file exists.

File: src/Drest/Mapping/Driver/JsonDriver.php [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+            unset($tmp['entity']);
+            $entities[$entity] = $tmp;
+        }
+        $this->classes = $entities;
+    }
+
+    /**
+     * 
+     */
+    public static function register(Configuration $config) {
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/JsonDriver.php]: Is this the same exact implementation from PhpDriver?

File: src/Drest/Mapping/Driver/JsonDriver.php [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+            self::$configuration_filename = $configuration_filename;
+        } else {
+            throw new \RuntimeException('You must set a configuration file path in index.php.');
+        }
+    }
+
+    /**
+     * Factory method for the Annotation Driver
+     *
+     * @param  Annotations\AnnotationReader $reader
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/JsonDriver.php]: Incorrect.

File: src/Drest/Mapping/Driver/JsonDriver.php [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+        if($configuration_filepath != null) {
+            self::$configuration_filepath = $configuration_filepath;
+            self::$configuration_filename = $configuration_filename;
+        } else {
+            throw new \RuntimeException('You must set a configuration file path in index.php.');
+        }
+    }
+
+    /**
+     * Factory method for the Annotation Driver
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/JsonDriver.php]: Bad docblock.

File: src/Drest/Mapping/Driver/JsonDriver.php [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+        return $this->classNames;
+    }
+
+    /**
+     * Load metadata for a class name
+     * @param  object|string         $class - Pass in either the class name, or an instance of that class
+     * @return Mapping\ClassMetaData $metaData - return null if metadata couldn't be populated from annotations
+     * @throws DrestException
+     */
+    public function loadMetadataForClass($class)
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/JsonDriver.php]: Why are you duplicating this function? Should be the same as in the PhpDriver. Same for others.

File: src/Drest/Mapping/Driver/PhpDriver.php [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+
+/**
+ * The PhpDriver reads a configuration file (config.php) rather than utilizing annotations.
+ */
+class PhpDriver implements DriverInterface
+{
+
+    static $configuration_filepath = null;
+    static $configuration_filename = null;
+    static $configuration_variable = 'resources';
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/PhpDriver.php]: This implementation should be changed. Instead, the config file should return and we can capture the return from the include.

File: src/Drest/Mapping/Driver/PhpDriver.php [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+     * @var array
+     */
+    protected $classes = [];
+
+    protected $resources = null; 
+
+    public function __construct()
+    {
+
+        $filename = self::$configuration_filepath . '\\' . self::$configuration_filename;
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/PhpDriver.php]: Use DIRECTORY_SEPARATOR here.

File: src/Drest/Mapping/Driver/PhpDriver.php [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+
+        file_exists($filename) && include($filename);
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/PhpDriver.php]: Throw an exception if the file doesn't exist.

File: src/Drest/Mapping/Driver/PhpDriver.php [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+
+        $var = self::$configuration_variable;
+        if(!isset($$var) || !is_array($$var)) {
+            throw new \RuntimeException('Invalid configuration file.');
+        }
+        $this->classes = $$var;
+
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/PhpDriver.php]: You need better code quality tools. Try and get PHPCS and PHPMD set up today.

File: src/Drest/Mapping/Driver/PhpDriver.php [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+     */
+    public static function register(Configuration $config) {
+        $configuration_filepath = $config->getAttribute('configFilePath');
+        $configuration_filename = $config->getAttribute('configFileName');
+
+        if($configuration_filepath != null) {
+            self::$configuration_filepath = $configuration_filepath;
+            self::$configuration_filename = $configuration_filename;
+        } else {
+            throw new \RuntimeException('You must set a configuration file path in index.php.');
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/PhpDriver.php]: The Config class might not necessarily be instantiated in a file called "index.php"

File: src/Drest/Mapping/Driver/PhpDriver.php [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+        if (is_array($resource['representations']))
+        {
+            $metadata->addRepresentations($resource['representations']);
+        }
+
+        $this->processRoutes($resource['routes'], $metadata);
+
+        $this->processMethods($resource, $metadata);
+
+
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/PhpDriver.php]: Should never have two empty lines in your code.

File: src/Drest/Mapping/Driver/PhpDriver.php [hunk 22]
```diff
@@ -0,0 +1,225 @@
...
+    }
+
+
+    /**
+     * Process the method
+     * @param $methods
+     * @param Mapping\ClassMetaData $metadata
+     * @throws DrestException
+     */
+    protected function processMethods($resource, Mapping\ClassMetaData $metadata)
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/PhpDriver.php]: If any of this code is copy/pasted from the AnnotationDriver class, you should create a base Driver class that implements DriverInterface and extend PhpDriver and AnnotationDriver from that class.

File: src/Drest/Mapping/Driver/YamlDriver.php [hunk 23]
```diff
@@ -0,0 +1,204 @@
+<?php
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/YamlDriver.php]: Loads of duplicated code here. Find a way to inject the parser into the class as a static property of the class in the register function. You should never have a new operator in a constructor.

File: src/Drest/Mapping/Driver/YamlDriver.php [hunk 24]
```diff
@@ -0,0 +1,204 @@
...
+ * The YamlDriver reads a configuration file (config.yaml) rather than utilizing annotations.
+ */
+class YamlDriver extends PhpDriver
+{
+
+    public function __construct()
+    {
+        $yaml = new Parser();
+        $filename = self::$configuration_filepath . '\\' . self::$configuration_filename;
+        echo $filename;
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/YamlDriver.php]: This is debug code.

File: src/Drest/Mapping/Driver/YamlDriver.php [hunk 25]
```diff
@@ -0,0 +1,204 @@
...
+        try 
+        {
+            $this->classes = $yaml->parse(file_get_contents($filename));
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/YamlDriver.php]: Does the file exist?

File: src/Drest/Mapping/Driver/YamlDriver.php [hunk 26]
```diff
@@ -0,0 +1,204 @@
...
+        } catch (ParseException $e) {
+            printf("Unable to parse the YAML string: %s", $e->getMessage());
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/YamlDriver.php]: Don't catch the exception. And definitely don't print an error from a library class. Ever.

File: src/Drest/Mapping/Driver/YamlDriver.php [hunk 27]
```diff
@@ -0,0 +1,204 @@
...
+        }
+    }
+
+    /**
+     * 
+     */
+    public static function register(Configuration $config) {
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/YamlDriver.php]: Don't copy/paste.

File: src/Drest/Mapping/Driver/YamlDriver.php [hunk 28]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    /**
+     * Factory method for the Annotation Driver
+     *
+     * @param  Annotations\AnnotationReader $reader
+     * @param  array|string                 $paths
+     * @return AnnotationDriver
+     */
+    public static function create($paths = [])
```
Comment (jdrich) [inline src/Drest/Mapping/Driver/YamlDriver.php]: Here as well.

File: tests/DrestTests/ConfigFiles/config.json [hunk 29]
```diff
@@ -0,0 +1,153 @@
+{
```
Comment (jdrich) [inline tests/DrestTests/ConfigFiles/config.json]: You've added a few empty files. You need to remove those from git.

Comment (jdrich) [inline tests/DrestTests/ConfigFiles/config.json]: These files should probably not exist in DrestTests. Are there other non-PHP files in DrestTests? I would put them in a directory under tests/

File: tests/DrestTests/ConfigFiles/config.php [hunk 30]
```diff
@@ -0,0 +1,26 @@
...
+        ['name' => 'get_user_profile', 'routePattern' => '/user/:id/profile', 'verbs' => ['GET'], 'expose' => ['profile']],
+        ['name' => 'get_user_numbers', 'routePattern' => '/user/:id/numbers', 'verbs' => ['GET'], 'expose' => ['phone_numbers']],
+        ['name' => 'post_user', 'routePattern' => '/user', 'verbs' => ['POST'], 'expose' => ['username', 'email_address', 'profile' => ['firstname', 'lastname'], 'phone_numbers' => ['number']], 'handle_call' => 'populatePost'],
+        ['name' => 'get_users', 'routePattern' => '/users', 'verbs' => ['GET'], 'collection' => 'true', 'expose' => ['username', 'email_address', 'profile', 'phone_numbers']],
+        ['name' => 'update_user', 'routePattern' => '/user/:id', 'routeConditions' => ['id' => '\d+'], 'verbs' => ['PUT', 'PATCH'], 'expose' => ['email_address', 'profile' => ['firstname', 'lastname']], 'handle_call' => 'patchUser'],
+        ['name' => 'delete_user', 'routePattern' => '/user/:id', 'verbs' => ['DELETE']],
+        ['name' => 'delete_users', 'routePattern' => '/users', 'collection' => 'true', 'verbs' => ['DELETE']]
+    ]
+];
```
Comment (jdrich) [inline tests/DrestTests/ConfigFiles/config.php]: If you return "$resources" at the end of this file, the new PhpDriver implementation for loading config will work.

File: tests/DrestTests/Mapping/AnnotationDriverTest.php [hunk 31]
```diff
@@ -0,0 +1,263 @@
...
+    
+    public function testMetadataConstruction()
+    {
+        $className = 'DrestTests\\Entities\\Typical\\User';
+        $cmd = new ClassMetaData(new \ReflectionClass($className));
+
+        $this->assertEquals($className, $cmd->getClassName());
+    }
+
+    public function testMetaDataCanAddRepresentation()
```
Comment (jdrich) [inline tests/DrestTests/Mapping/AnnotationDriverTest.php]: This is not a test of the AnnotationDriver. Should be moved back into the ClassMetaData tests.

File: tests/DrestTests/Mapping/AnnotationDriverTest.php [hunk 32]
```diff
@@ -0,0 +1,263 @@
...
+        $className = 'DrestTests\\Entities\\Typical\\User';
+        $cmd = new ClassMetaData(new \ReflectionClass($className));
+
+        $rep = new Json();
+        $cmd->addRepresentation($rep);
+
+        $this->assertContains($rep, $cmd->getRepresentations());
+    }
+
+    public function testMetadataCanAddStringRepresentation()
```
Comment (jdrich) [inline tests/DrestTests/Mapping/AnnotationDriverTest.php]: Also not an AnnotationDriver test.

File: tests/DrestTests/Mapping/AnnotationDriverTest.php [hunk 33]
```diff
@@ -0,0 +1,263 @@
...
+        $rep = 'SomeClass';
+        $cmd->addRepresentation($rep);
+
+        $this->assertContains($rep, $cmd->getRepresentations());
+    }
+
+    /**
+     * @expectedException \DrestCommon\Representation\RepresentationException
+     */
+    public function testMetadataInvalidRepresentationObject()
```
Comment (jdrich) [inline tests/DrestTests/Mapping/AnnotationDriverTest.php]: Nope.

File: tests/DrestTests/Mapping/AnnotationDriverTest.php [hunk 34]
```diff
@@ -0,0 +1,263 @@
...
+        $cmd = new ClassMetaData(new \ReflectionClass($className));
+
+        $rep = new \StdClass();
+        $cmd->addRepresentation($rep);
+    }
+
+    /**
+     * @expectedException \DrestCommon\Representation\RepresentationException
+     */
+    public function testMetadataRepresentationCanNotBeAnArray()
```
Comment (jdrich) [inline tests/DrestTests/Mapping/AnnotationDriverTest.php]: Nope.

File: tests/DrestTests/Mapping/AnnotationDriverTest.php [hunk 35]
```diff
@@ -0,0 +1,263 @@
...
+        $cmd = new ClassMetaData(new \ReflectionClass($className));
+
+        $rep = array('a', 'b');
+        $cmd->addRepresentation($rep);
+    }
+
+    /**
+     * @expectedException \DrestCommon\Representation\RepresentationException
+     */
+    public function testMetadataRepresentationCanNotBeAnInteger()
```
Comment (jdrich) [inline tests/DrestTests/Mapping/AnnotationDriverTest.php]: Nope.

File: tests/DrestTests/Mapping/JsonDriverTest.php [hunk 36]
```diff
@@ -0,0 +1,50 @@
...
+
+use Drest\Configuration;
+use Drest\Mapping\ClassMetaData;
+use Drest\Mapping\MetadataFactory;
+use Drest\Mapping\Driver\JsonDriver;
+
+class JsonDriverTest extends \PHPUnit_Framework_TestCase
+{
+
+    // test register function with null config file
```
Comment (jdrich) [inline tests/DrestTests/Mapping/JsonDriverTest.php]: This isn't a proper docblock. Also, the comment is actually not correct - you are testing an invalid configuration here, not a null config file.

File: tests/DrestTests/Mapping/JsonDriverTest.php [hunk 37]
```diff
@@ -0,0 +1,50 @@
...
+    public function testRegisterWithNoConfig() {
+        $this->setExpectedException('RuntimeException');
+        $configuration = new Configuration();
+        JsonDriver::register($configuration);
+    }
+
+    // test register function with a non-existing directory for the configuration file
+    public function testRegisterDirNotExists() {
+        $configuration = new Configuration();
```
Comment (jdrich) [inline tests/DrestTests/Mapping/JsonDriverTest.php]: You're not testing anything here. If the test doesn't have an assertion or expect an exception, nothing is actually being tested.

File: tests/DrestTests/Mapping/JsonDriverTest.php [hunk 38]
```diff
@@ -0,0 +1,50 @@
...
+        $configuration->setAttribute('configFilePath', '/BadConfigFiles');
+        JsonDriver::register($configuration);     
+    }
+
+    // ensuring that the configuration file exists
+    public function testRegisterFileNotExists() {
+        $configuration = new Configuration();
```
Comment (jdrich) [inline tests/DrestTests/Mapping/JsonDriverTest.php]: Same here and other places.

File: tests/DrestTests/Mapping/JsonDriverTest.php [hunk 39]
```diff
@@ -0,0 +1,50 @@
...
+
+    public function testRegisterFileInvalid() {
+        $this->setExpectedException('RuntimeException');
+        $configuration = new Configuration();
+        JsonDriver::register($configuration);
+        $configuration->setAttribute('configFilePath', '$');
+    }
+
+    public function testRegisterFileValid() {
+        $configFilePath = '../ConfigFiles/';
```
Comment (jdrich) [inline tests/DrestTests/Mapping/JsonDriverTest.php]: No assertion -> No Test.

File: tests/DrestTests/Mapping/PhpDriverTest.php [hunk 40]
```diff
@@ -0,0 +1,314 @@
+<?php
+
+namespace DrestTests\Mapping;
+
+use Drest\Configuration;
```
Comment (jdrich) [inline tests/DrestTests/Mapping/PhpDriverTest.php]: Make sure you're actually using everything here.

File: tests/DrestTests/Mapping/PhpDriverTest.php [hunk 41]
```diff
@@ -0,0 +1,314 @@
...
+        PhpDriver::register($configuration);
+    }
+
+    /* Below are the tests ported over from AnnotationDriver. */
+
+    // Adding a representation to the config
+    public function testMetaDataCanAddRepresentation()
+    {
+        include($this->createGoodTmpFile());
+        if(in_array('representations', $resources['\Entities\User']) && !in_array('Json', $resources['\Entities\User']['representations'])) {
```
Comment (jdrich) [inline tests/DrestTests/Mapping/PhpDriverTest.php]: You're not actually pushing a representation here. Also, the include does nothing. Try and comprehend what the other tests you're looking at are actually doing.

File: tests/DrestTests/Mapping/PhpDriverTest.php [hunk 42]
```diff
@@ -0,0 +1,314 @@
...
+    'routes' => [
+        ['name' => 'get_user', 'routePattern' => '/user/:id', 'routeConditions' => ['id' => '\d+'], 'verbs' => ['GET'], 'action' => 'Action\Custom'],
+        ['name' => 'get_user', 'routePattern' => '/user/:id', 'routeConditions' => ['id' => '\d+'], 'verbs' => ['GET'], 'action' => 'Action\Custom'],
+    ]
+];
+HEREDOC;
+        $tmp = $this->createCustomTmpFile($file_contents);
+        $metadataFactory = new MetadataFactory(
+            \Drest\Mapping\Driver\PhpDriver::create(
+                array($tmp)
```
Comment (jdrich) [inline tests/DrestTests/Mapping/PhpDriverTest.php]: This doesn't actually do what you're thinking it does. You're also not registering the PhpDriver with a config so how can this be doing anything at all?

**Comment by jdrich** [comment]
Added a whole slew of comments. You really need to reassess how you look at other people's code - this code isn't very complex, but I think you're assuming too quickly that you know what the existing tests are testing. Rejecting for now. Fix my notes and resubmit.
